sui/deploy-contract.js

```diff
@@ -23,49 +54,179 @@ async function processCommand(contractName, config, chain, options) {
+                target: `${packageId}::bytes32::new`,
+                arguments: [tx.pure(arrayify(domainSeparator))],
+            });
+
+            tx.moveCall({
+                target: `${packageId}::gateway::setup`,
+                arguments: [
+                    tx.object(creatorCap),
+                    tx.pure.address(operator),
+                    separator,
+                    tx.pure.u64(minimumRotationDelay),
+                    tx.pure.u64(options.previousSigners),
+                    tx.pure(bcs.vector(bcs.u8()).serialize(encodedSigners).toBytes()),
+                    tx.object('0x6'),
+                ],
+            });
+
+            if (policy !== '0') {
+                const upgradeType = policy === '128' ? 'only_additive_upgrades' : 'only_dep_upgrades';
+
+                tx.moveCall({
+                    target: `0x2::package::${upgradeType}`,
+                    arguments: [tx.object(upgradeCap)],
+                });
+            }
+
+            const result = await broadcast(client, keypair, tx);
+
+            printInfo('Setup transaction digest: ', result.digest);
+
+            const [gateway] = getObjectIdsByObjectTypes(result, [`${packageId}::gateway::Gateway`]);
+
+            contractConfig.objects = {
+                gateway,
+                relayerDiscovery,
+                upgradeCap,
+            };
+            contractConfig.domainSeparator = domainSeparator;
+            contractConfig.operator = operator;
+            contractConfig.minimumRotationDelay = minimumRotationDelay;
+            break;
+        }
+
+        default: {
             throw new Error(`${contractName} is not supported.`);
+        }
     }
 
     printInfo(`${contractName} deployed`, JSON.stringify(contractConfig, null, 2));
 }
 
-async function mainProcessor(contractName, options, processor) {
+async function upgrade(contractName, policy, config, chain, options) {
+    const [keypair, client] = getWallet(chain, options);
+    const { packageDependencies } = options;
+    options.policy = policy;
+
+    await printWalletInfo(keypair, client, chain, options);
+
+    if (!chain.contracts[contractName]) {
+        throw new Error(`Cannot find specified contract: ${contractName}`);
+    }
+
+    const contractsConfig = chain.contracts;
+    const packageConfig = contractsConfig?.[contractName];
+
+    validateParameters({ isNonEmptyString: { contractName } });
+
+    if (packageDependencies) {
+        for (const dependencies of packageDependencies) {
+            const packageId = contractsConfig[dependencies]?.address;
+            updateMoveToml(dependencies, packageId);
+        }
+    }
+
+    const builder = new TxBuilder(client);
+    await upgradePackage(client, keypair, contractName, packageConfig, builder, options);
+}
+
+async function mainProcessor(args, options, processor) {
     const config = loadSuiConfig(options.env);
-    await processor(contractName, config, config.sui, options);
```

Comment:
perform contractName validation here, since it'll be provided for any processor method